modules/programs/java.nix

```
@@ -0,0 +1,73 @@
+# This module provides JAVA_HOME, with a different way to install java
+# locally.
+# This module is modified from the NixOS module `programs.java`
+
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let cfg = config.programs.java;
+
+in {
+  meta.maintainers = with maintainers; [ ShamrockLee ];
+
+  options = {
+
+    programs.java = {
+
+      enable = mkEnableOption "java" // {
+        description = ''
+          Install and setup the Java development kit.
+          <note>
+          <para>This adds JAVA_HOME to the user's home environment, by extracting the variable
+            from the jdk's setup-hook and put inside home.sessionVariables.
+            It is equivalent to starting a shell
+            through 'nix-shell -p jdk', or roughly the following Home Manager
+            configuration:
+          </para>
+          <programlisting>
+            home.sessionVariables.JAVA_HOME = ''${pkgs.jdk.home}/lib/openjdk;
+            home.packages = [ pkgs.jdk ];
+          </programlisting>
+          <para>This module make use of <literal>home.sessionVariables</literal>.
+            To use it in graphical sessions, set
+          </para>
+          <programlisting>
+            xsession.enable = true;
+          </programlisting>
+          <para>or source <literal>/home/shamrock/.nix-profile/etc/profile.d/hm-session-vars.sh</literal>
+            when you need the variables including <literal>JAVA_HOME</literal>.
+          </para>
+          </note>
+        '';
+      };
+
+      package = mkOption {
+        default = pkgs.jdk;
+        defaultText = "pkgs.jdk";
+        description = ''
+          Java package to install. Typical values are pkgs.jdk or pkgs.jre.
+        '';
+        type = types.package;
+      };
+
+    };
+
+  };
+
+  config = mkIf cfg.enable {
+
+    home.packages = [ cfg.package ];
+
+    home.sessionVariables = {
+      JAVA_HOME = pkgs.lib.strings.fileContents "${
+          pkgs.runCommandNoCC "java-home" { } ''
+            source "${cfg.package}/nix-support/setup-hook"
+            mkdir -p $out
+            echo "$JAVA_HOME" > $out/java-home
```

Maybe
```suggestion
            echo "$JAVA_HOME" > $out
```
then it should be sufficient to have
```nix
fileContents (pkgs.runCommandLocal "java-home" { } ''
  …
'')
```